worlds/doom_ii/__init__.py

```diff
@@ -0,0 +1,268 @@
+from worlds.AutoWorld import WebWorld, World
+from . import Items, Locations, Maps, Options, Regions, Rules
+
+logger = logging.getLogger("DOOM II")
+
+DOOM_TYPE_LEVEL_COMPLETE = -2
+DOOM_TYPE_COMPUTER_AREA_MAP = 2026
+
+
+class DOOM2Location(Location):
+    game: str = "DOOM II"
+
+
+class DOOM2Item(Item):
+    game: str = "DOOM II"
+
+
+class DOOM2Web(WebWorld):
+    tutorials = [Tutorial(
+        "Multiworld Setup Guide",
+        "A guide to setting up the DOOM II randomizer connected to an Archipelago Multiworld",
+        "English",
+        "setup_en.md",
+        "setup/en",
+        ["Daivuk"]
+    )]
+    theme = "dirt"
+
+
+class DOOM2World(World):
+    """
+    Doom II, also known as Doom II: Hell on Earth, is a first-person shooter game by id Software.
+    It was released for MS-DOS in 1994.
+    Compared to its predecessor, Doom II features larger levels, new enemies, a new "super shotgun" weapon
+    """
+    option_definitions = Options.options
+    game = "DOOM II"
+    web = DOOM2Web()
+    data_version = 3
+    required_client_version = (0, 3, 9)
+
+    item_name_to_id = {data["name"]: item_id for item_id, data in Items.item_table.items()}
+    item_name_groups = Items.item_name_groups
+
+    location_name_to_id = {data["name"]: loc_id for loc_id, data in Locations.location_table.items()}
+    location_name_groups = Locations.location_name_groups
+
+    starting_level_for_episode: List[str] = [
+        "Entryway (MAP01)",
+        "The Factory (MAP12)",
+        "Nirvana (MAP21)"
+    ]
+
+    # Item ratio that scales depending on episode count. These are the ratio for 3 episode. In DOOM1.
+    # The ratio have been tweaked seem, and feel good.
+    items_ratio: Dict[str, float] = {
+        "Armor": 41,
+        "Mega Armor": 25,
+        "Berserk": 12,
+        "Invulnerability": 10,
+        "Partial invisibility": 18,
+        "Supercharge": 28,
+        "Medikit": 15,
+        "Box of bullets": 13,
+        "Box of rockets": 13,
+        "Box of shotgun shells": 13,
+        "Energy cell pack": 10
+    }
+
+    def __init__(self, world: MultiWorld, player: int):
+        self.included_episodes = [1, 1, 1, 0]
+        self.location_count = 0
+
+        super().__init__(world, player)
+
+    def get_episode_count(self):
+        count = 0
+        for i in range(3): # Don't include 4th, those are secret levels they are additive
+            count += self.included_episodes[i]
+        return count
+
+    def generate_early(self):
+        # Cache which episodes are included
+        for i in range(3):
+            self.included_episodes[i] = getattr(self.multiworld, f"episode{i + 1}")[self.player].value
+
+        # If no episodes selected, select Episode 1
+        if self.get_episode_count() == 0:
+            self.included_episodes[0] = 1
+
+        # 4th episode are secret levels
+        self.included_episodes[3] = getattr(self.multiworld, "episode4")[self.player].value
```

Comment:
Since `get_episode_count` doesn't include the 4th episode anyway, this can be slimmer.

```suggestion
        for i in range(4):
            self.included_episodes[i] = getattr(self.options, f"episode{i + 1}").value

        # If no episodes selected, select Episode 1
        if self.get_episode_count() == 0:
            self.included_episodes[0] = 1
```